File: v1-migration-backend/src/routes/internal.ts
import { Hono } from "hono";
import type { AppEnv } from "../env";
import { createDb } from "../db/client";
import { requireAuth } from "../middleware/auth";
import { getImageUrl } from "../storage";

const internal = new Hono<AppEnv>();

internal.post("/inference-callback", async (c) => {
  const payload = await c.req.json<{
    batch_id?: string;
    results?: Array<{
      photo_id?: string;
      faces?: Array<{
        bbox?: Record<string, number>;
        embedding?: number[];
        confidence?: number;
      }>;
    }>;
  }>();

  const results = payload.results || [];
  const batchId = payload.batch_id;
  const similarityThreshold = Number.parseFloat(c.env.SIMILARITY_THRESHOLD || "0.35");
  const maxDistance = 1.0 - similarityThreshold;

  console.log(
    `Received inference callback for batch: ${batchId}. Images count: ${results.length}`,
  );

  const sql = createDb(c.env);
  try {
    const countRows = await sql`SELECT COUNT(*)::int AS count FROM face_embeddings`;
    console.log(
      `DEBUG: total face_embeddings visible to this session: ${countRows[0].count}`,
    );

    for (const item of results) {
      const photoId = item.photo_id;
      if (!photoId) continue;

      const photoRows = await sql`
        SELECT id, owner_id FROM photos WHERE id = ${photoId}::uuid
      `;
      if (photoRows.length === 0) {
        console.log(`Warning: Photo ${photoId} not found in database.`);
        continue;
      }

      const photo = photoRows[0];
      const faces = item.faces || [];

      for (const face of faces) {
        const bbox = face.bbox || {};
        const embedding = face.embedding || [];
        const confidence = face.confidence ?? 1.0;

        if (embedding.length !== 512) continue;

        const embeddingLiteral = JSON.stringify(embedding);
        const matchRows = await sql`
          SELECT user_id, (embedding <=> ${embeddingLiteral}::vector) AS distance
          FROM face_embeddings
          WHERE embedding <=> ${embeddingLiteral}::vector <= ${maxDistance}
          ORDER BY distance ASC
          LIMIT 1
        `;

        if (matchRows.length > 0) {
          const matchedUserId = matchRows[0].user_id as string;
          const distance = matchRows[0].distance as number;
          console.log(
            `Matched face in photo ${photoId} to user ${matchedUserId} (distance: ${distance.toFixed(4)})`,
          );

          const xmin = bbox.xmin ?? bbox.x;
          const ymin = bbox.ymin ?? bbox.y;
          const xmax = bbox.xmax;
          const yminVal = ymin ?? 0;
          const xminVal = xmin ?? 0;
          const width =
            xmax !== undefined && xmin !== undefined ? xmax - xmin : 0.1;
          const height =
            bbox.ymax !== undefined && ymin !== undefined
              ? bbox.ymax - ymin
              : 0.1;

          await sql`
            INSERT INTO photo_tags (
              photo_id, user_id, bbox_x, bbox_y, bbox_width, bbox_height,
              confidence, source
            )
            VALUES (
              ${photoId}::uuid,
              ${matchedUserId}::uuid,
              ${xminVal},
              ${yminVal},
              ${width},
              ${height},
              ${confidence},
              'auto'
            )
          `;

          await sql`
            INSERT INTO notifications (user_id, type, photo_id, actor_user_id, is_read)
            VALUES (
              ${matchedUserId}::uuid,
              'tagged_in_photo',
              ${photoId}::uuid,
              ${photo.owner_id}::uuid,
              false
            )
          `;
        } else {
          const closestRows = await sql`
            SELECT user_id, (embedding <=> ${embeddingLiteral}::vector) AS distance
            FROM face_embeddings
            ORDER BY distance ASC
            LIMIT 1
          `;
          if (closestRows.length > 0) {
            const closestSim = 1.0 - (closestRows[0].distance as number);
            console.log(
              `No match found for face in photo ${photoId} (closest was user ${closestRows[0].user_id}, similarity: ${closestSim.toFixed(4)})`,
            );
          } else {
            console.log(
              `No match found for face in photo ${photoId} (no embeddings in DB at all)`,
            );
          }
        }
      }

      await sql`
        UPDATE photos
        SET status = 'processed',
            processed_at = NOW(),
            batch_id = ${batchId ?? null}::uuid
        WHERE id = ${photoId}::uuid
      `;
    }

    return c.json({ acknowledged: true });
  } finally {
    await sql.end({ timeout: 5 });
  }
});

export default internal;

File: v1-migration-backend/src/routes/internal.py
"""Internal inference callback."""

from __future__ import annotations

from db import create_db, embedding_literal
from http_util import json_response, parse_json


async def handle_internal(env, method: str, subpath: str, request, _user_id: str | None):
    if method == "POST" and subpath == "inference-callback":
        return await inference_callback(env, request)
    return json_response({"detail": "Not found"}, 404)


async def inference_callback(env, request):
    payload = await parse_json(request)
    results = payload.get("results") or []
    batch_id = payload.get("batch_id")
    similarity_threshold = float(getattr(env, "SIMILARITY_THRESHOLD", None) or "0.35")
    max_distance = 1.0 - similarity_threshold

    print(f"Received inference callback for batch: {batch_id}. Images count: {len(results)}")

    db = create_db(env.DATABASE_URL)
    try:
        count_row = db.fetchone("SELECT COUNT(*)::int AS count FROM face_embeddings")
        print(f"DEBUG: total face_embeddings visible to this session: {count_row['count']}")

        for item in results:
            photo_id = item.get("photo_id")
            if not photo_id:
                continue

            photo = db.fetchone(
                "SELECT id, owner_id FROM photos WHERE id = %s::uuid",
                (photo_id,),
            )
            if not photo:
                print(f"Warning: Photo {photo_id} not found in database.")
                continue

            for face in item.get("faces") or []:
                bbox = face.get("bbox") or {}
                embedding = face.get("embedding") or []
                confidence = face.get("confidence", 1.0)
                if len(embedding) != 512:
                    continue

                emb_lit = embedding_literal(embedding)
                match = db.fetchone(
                    """
                    SELECT user_id, (embedding <=> %s::vector) AS distance
                    FROM face_embeddings
                    WHERE embedding <=> %s::vector <= %s
                    ORDER BY distance ASC
                    LIMIT 1
                    """,
                    (emb_lit, emb_lit, max_distance),
                )

                if match:
                    matched_user_id = str(match["user_id"])
                    distance = float(match["distance"])
                    print(
                        f"Matched face in photo {photo_id} to user {matched_user_id} "
                        f"(distance: {distance:.4f})"
                    )
                    xmin = bbox.get("xmin", bbox.get("x"))
                    ymin = bbox.get("ymin", bbox.get("y"))
                    xmax = bbox.get("xmax")
                    ymax = bbox.get("ymax")
                    xmin_val = xmin if xmin is not None else 0
                    ymin_val = ymin if ymin is not None else 0
                    width = (xmax - xmin) if xmax is not None and xmin is not None else 0.1
                    height = (ymax - ymin) if ymax is not None and ymin is not None else 0.1

                    db.execute(
                        """
                        INSERT INTO photo_tags (
                            photo_id, user_id, bbox_x, bbox_y, bbox_width, bbox_height,
                            confidence, source
                        )
                        VALUES (%s::uuid, %s::uuid, %s, %s, %s, %s, %s, 'auto')
                        """,
                        (photo_id, matched_user_id, xmin_val, ymin_val, width, height, confidence),
                    )
                    db.execute(
                        """
                        INSERT INTO notifications (user_id, type, photo_id, actor_user_id, is_read)
                        VALUES (%s::uuid, 'tagged_in_photo', %s::uuid, %s::uuid, false)
                        """,
                        (matched_user_id, photo_id, str(photo["owner_id"])),
                    )
                else:
                    closest = db.fetchone(
                        """
                        SELECT user_id, (embedding <=> %s::vector) AS distance
                        FROM face_embeddings
                        ORDER BY distance ASC
                        LIMIT 1
                        """,
                        (emb_lit,),
                    )
                    if closest:
                        closest_sim = 1.0 - float(closest["distance"])
                        print(
                            f"No match found for face in photo {photo_id} "
                            f"(closest was user {closest['user_id']}, similarity: {closest_sim:.4f})"
                        )
                    else:
                        print(f"No match found for face in photo {photo_id} (no embeddings in DB at all)")

            db.execute(
                """
                UPDATE photos
                SET status = 'processed', processed_at = NOW(), batch_id = %s::uuid
                WHERE id = %s::uuid
                """,
                (batch_id, photo_id),
            )

        return json_response({"acknowledged": True})
    finally:
        db.close()
